ipa-core/src/protocol/ipa_prf/malicious_security/hashing.rs

```diff
@@ -0,0 +1,171 @@
+    // however, deserialize might fail for some fields so we use `from_random_128` instead
+    // this results in at most 128 bits of security/collision probability rather than 256 bits as offered by `Sha256`
+    // for field elements of size less than 129 bits, this does not make a difference
+    F::from_random_u128(u128::from_le_bytes(
+        sha.finalize()[0..16].try_into().unwrap(),
+    ))
+}
+
+#[cfg(all(test, unit_test))]
+mod test {
+    use rand::{thread_rng, Rng};
+
+    use super::compute_hash;
+    use crate::{
+        ff::{Fp31, Fp32BitPrime},
+        protocol::ipa_prf::malicious_security::hashing::hash_to_field,
+    };
+
+    #[test]
+    fn hash_changes() {
+        const LIST_LENGTH: usize = 5;
+
+        let mut rng = thread_rng();
+
+        let mut list: Vec<Fp31> = Vec::with_capacity(LIST_LENGTH);
+        for _ in 0..LIST_LENGTH {
+            list.push(rng.gen::<Fp31>());
+        }
+        let hash_1 = compute_hash(&list);
+
+        // modify one, randomly selected element in the list
+        let random_index = rng.gen::<usize>() % LIST_LENGTH;
+        let mut different_field_element = list[random_index];
+        while different_field_element == list[random_index] {
+            different_field_element = rng.gen::<Fp31>();
+        }
+        list[random_index] = different_field_element;
+
+        let hash_2 = compute_hash(&list);
+
+        assert_ne!(
+            hash_1, hash_2,
+            "The hash should change if the input is different"
+        );
+
+        // swapping two elements should change the hash
+        let mut index_1 = 0;
+        let mut index_2 = 0;
+        // There is a 1 in 31 chance that these two elements are the exact same value.
+        // To make sure this doesn't become a flaky test, let's just pick two different
+        // elements to swap when that happens.
+        // This will be an infinite loop if all elements are the same. The chances of that
+        // are (1 / 31) ^ (LIST_LENGTH - 1)
+        // which is 1 in 923,521 when LIST_LENGTH is 5. I'm OK with that.
+        while list[index_1] == list[index_2] {
+            index_1 = rng.gen_range(0..LIST_LENGTH);
+            index_2 = (index_1 + rng.gen_range(1..LIST_LENGTH)) % LIST_LENGTH;
+        }
+        list.swap(index_1, index_2);
+
+        let hash_3 = compute_hash(&list);
+
+        assert_ne!(
+            hash_2, hash_3,
+            "The hash should change if two elements are swapped"
+        );
+    }
+
+    #[test]
+    fn field_element_changes() {
+        const LIST_LENGTH: usize = 5;
+
+        let mut rng = thread_rng();
+
+        let mut left = Vec::with_capacity(LIST_LENGTH);
+        let mut right = Vec::with_capacity(LIST_LENGTH);
+        for _ in 0..LIST_LENGTH {
+            left.push(rng.gen::<Fp32BitPrime>());
+            right.push(rng.gen::<Fp32BitPrime>());
+        }
+        let r1: Fp32BitPrime = hash_to_field(&compute_hash(&left), &compute_hash(&right));
```

Comment:
Speaking of collisions... 

I generated this test case in a Google spreadsheet and copied all the values in here, including the random values chosen for "r". Well, now that the code actually hashes the shares of the proof in order to select "r", I had to iterate through many different secret-sharings in order to find one that would result in the exact same value of "r" being selected as I had before. It didn't take that long since I'm using Fp31, but yeah...

Reply:
Yes, we definitely need that there is enough entropy in the input, i.e. we cannot use proofs that are too short. For 128 bits of security we need an input length of at least 128 bit. Since the proof length is 2 lambda - 1, we need that the size of field elements is at least 128/(2 lambda -1) bits.